Request Streaming upload (#33693)
dotnet/AspNetCore.Docs: aspnetcore/blazor/call-web-api.md

:::moniker-end

:::moniker range=">= aspnetcore-9.0"

## Client-side request streaming

> For Chromium-based browsers (for example, Google Chrome and Microsoft Edge) using the HTTP/2 protocol, HTTPS, and [CORS](xref:security/cors), client-side Blazor uses [Streams API](https://developer.mozilla.org/docs/Web/API/Streams_API) to permit [request streaming](https://developer.chrome.com/docs/capabilities/web-apis/fetch-streaming-requests).

Reviewer: This makes it sound like CORS is always required for request streaming, even for same origin requests. Is this really correct? I suspect that CORS is only required for cross origin requests, just like with any other fetch request.

Author: ```suggestion
For Chromium-based browsers (for example, Google Chrome and Microsoft Edge) using the HTTP/2 protocol, and HTTPS, client-side Blazor uses [Streams API](https://developer.mozilla.org/docs/Web/API/Streams_API) to permit [request streaming](https://developer.chrome.com/docs/capabilities/web-apis/fetch-streaming-requests).
```

To enable request streaming, set <xref:Microsoft.AspNetCore.Components.WebAssembly.Http.WebAssemblyHttpRequestMessageExtensions.SetBrowserRequestStreamingEnabled%2A> to `true` on the <xref:System.Net.Http.HttpRequestMessage>.

In the following file upload example:

* `content` is the file's <xref:System.Net.Http.HttpContent>.
* `/Filesave` is the web API endpoint.
* `Http` is the <xref:System.Net.Http.HttpClient>.

```csharp
var request = new HttpRequestMessage(HttpMethod.Post, "/Filesave");
request.SetBrowserRequestStreamingEnabled(true);
request.Content = content;

var response = await Http.SendAsync(request);
```

Streaming requests:

* Require HTTPS protocol and don't work on HTTP/1.x.

> * Include a body but not a `Content-Length` header. [CORS](xref:security/cors) is required, and a CORS preflight request is always issued.

Reviewer: I see the CORS statement here is from https://developer.chrome.com/docs/capabilities/web-apis/fetch-streaming-requests, but again I suspect CORS is only really required for cross origin requests. I mean, does it work if you don't do anything to set up CORS on the server? If yes, then I suggest:

```suggestion
* Include a body but not a `Content-Length` header. [CORS](xref:security/cors) with a preflight reqeust is required for cross-origin streaming requests.
```

Author: Yes, @danroth27, that's the source. The PU referred to that Chrome doc.

@pavelsavara @campersau ... WRT Request Streaming Upload, the remarks are ...

> Requires CORS and triggers a preflight
>
> Streaming requests have a body, but don't have a `Content-Length` header. That's a new kind of request, so CORS is required, and these requests always trigger a preflight.
>
> Streaming `no-cors` requests are not allowed.

Cross-ref: https://developer.chrome.com/docs/capabilities/web-apis/fetch-streaming-requests#requires_cors_and_triggers_a_preflight

It's listed under "Restrictions" and makes it sound like a preflight will be issued. I take it now that that's ***incorrect*** for a same-origin request ... *correct?* 😄

Reviewer 2: This is what is listed in the chrome status feature description:

> - Only 'cors' and 'same-origin' requests allow streaming upload. You can't use streaming upload with 'navigate' and 'no-cors' requests.
> - This feature cannot be used with HTTP/1.x. If the server doesn't support HTTP/2 or HTTP/3, the request fails. This is for some compatibility concerns. See whatwg/fetch#966 for the past discussions.

https://chromestatus.com/feature/5274139738767360

So to me it looks like `no-cors` is not allowed but `cors` and `same-origin` are. I also tested it: same origin requests don't trigger preflight requests.

Reviewer: Yup, that all makes sense. The Chrome blog post is referring to the allowed [request mode](https://developer.mozilla.org/en-US/docs/Web/API/Request/mode). But when we talk about CORS being required in ASP.NET Core, that makes it sound like you have to enable CORS on the server, which isn't really the case.
